### PR TITLE
Add Tag index page

### PR DESCRIPTION
**Pull request by creeperyang**
Add tag index page generation for `hexo-generator-tag`.

After generation, it would be like:

```
public--
       |--tags--
                |----index.html (added)
                |----xxx
```

So like `/tag` menu will be supported.

### CODE REVIEW

**Comment by Xuanwo** [comment]
will it override my tags new?

**Comment by creeperyang** [comment]
No, just append to your generated pages or do nothing, controlled via tag_generator option:

```yaml
tag_generator:
  per_page: 10
  # append index page or do nothing
  enable_index_page: true/false
  # if enable_index_page: true, config template to use. ['tagIndex', 'tag', 'archive', 'index']
  template: 'tagIndex' 
```

**Comment by tommy351** [comment]
LGTM :+1:

File: lib/generator.js [hunk 1]
```diff
@@ -22,4 +25,32 @@ module.exports = function(locals){
...
+  // generate tag index page, usually /tags/index.html
+  if(tagConfig.enable_index_page) {
+    tagDir = config.tag_dir;
+    if (tagDir[tagDir.length - 1] !== '/') {
+     tagDir += '/';
+    }
+    pages.push({
+      path: tagDir,
+      layout: [tagConfig.template ? tagConfig.template : 
+        'tagIndex', 'tag', 'archive', 'index'],
```
Comment (tommy351) [inline lib/generator.js]: You can use:

``` js
[].concat(tagConfig.template, ['tagIndex', 'tag', 'archive', 'index']);
```

In this way, users will still have the template fallback when the template does not exist. Also I think `tag-index` would be better than `tagIndex`.

Comment (creeperyang) [inline lib/generator.js]: Your suggestion seems to be better. :)

**Comment by creeperyang** [comment]
@tommy351 I make a change after look through the source code of `archive` and `category` generator. And now, there is only one config option `enable_index_page: true/false`.

For template, it's of a particular sequence `['tag-index', 'tag', 'archive', 'index']`.

**Comment by borisschapira** [comment]
+1

**Comment by borisschapira** [comment]
@creeperyang I would like to use it but I can't manage to understand how to do it. Could you please explain ?

I've set this in configuration :

```
tag_generator:
  enable_index_page: true
  # if enable_index_page: true, config template to use. ['tagIndex', 'tag', 'archive', 'index']
  template: 'tagIndex' 
```

And added a `tagIndex.ejs` to my theme's supported layout. Anything else ?

**Comment by creeperyang** [comment]
@borisschapira  A little change done and now it's stable(merged). So currently the config is like:

```yaml
tag_generator:
  per_page: 10
  # append index page or do nothing ------ the only config option
  enable_index_page: true/false
  # template: 'tagIndex' ------ removed, default to 'tag-index', so use 'tag-index.ejs' instead.
```

And the variables inserted into the template(`tag-index`) are like:

```js
{
        base: tagDir,
        total: 1,
        current: 1,
        current_url: tagDir,
        posts: locals.posts, // all posts
        prev: 0,
        prev_link: '',
        next: 0,
        next_link: '',
        tags: tags // all tags
 }
```

Last, I check via ` npm info hexo-generator-tag` and @tommy351 hasn't made a new release. So you have to clone the repo if you want to use now.

**Comment by borisschapira** [comment]
Thanks for the answer, it works.

Maybe you should Pull-Request some evolution to the documentation with all of that info.

> Last, I check via npm info hexo-generator-tag and @tommy351 hasn't made a new release. So you have to clone the repo if you want to use now.

:hand: You can reference the repo like that to get last commit on master :smile: 
```
  "dependencies": {
    ...
    "hexo-generator-tag": "hexojs/hexo-generator-tag",
    ...
}
```

**Comment by borisschapira** [comment]
@creeperyang Do you plan to do the same with [hexojs/hexo-generator-category](https://github.com/hexojs/hexo-generator-category) ? If yes, maybe you should harmonize the structure of data passed to `tag-index` to allow the use of a *master* template (`taxonomy-index` ?) allowing the listing of tags or categories, regarding the route.

**Comment by borisschapira** [comment]
@creeperyang : another note : you list all the tags, including the tags of non-published drafts (therefor, the tag page does not necessarily exists). I suggest you filter on `tag.length`.

**Comment by borisschapira** [comment]
@creeperyang : and maybe also add a configuration for desired order (alpha, length, reverse...).

**Comment by creeperyang** [comment]
@borisschapira Good suggestions. But currentlly I have no time to do it. And I use `issues` to write blog now. Maybe it's a common demand to generate index page for tag and category, I'll fulfill it if the maintainers wont do it.

**Pull request merged**